src/field_theory/finite.lean

```diff
@@ -1,138 +1,247 @@
   congr; simp [set.ext_iff, classical.em]
 end
 
-instance : is_cyclic (units α) :=
-by haveI := classical.dec_eq α;
-haveI := set_fintype (@set.univ (units α)); exact
-let ⟨g, hg⟩ := is_cyclic.exists_generator (@set.univ (units α)) in
+/-- The units of a finite field form a cyclic group. -/
+instance : is_cyclic (units K) :=
+by haveI := classical.dec_eq K;
+haveI := set_fintype (@set.univ (units K)); exact
+let ⟨g, hg⟩ := is_cyclic.exists_generator (@set.univ (units K)) in
 ⟨⟨g, λ x, let ⟨n, hn⟩ := hg ⟨x, trivial⟩ in ⟨n, by rw [← is_subgroup.coe_gpow, hn]; refl⟩⟩⟩
 
 lemma prod_univ_units_id_eq_neg_one :
-  univ.prod (λ x, x) = (-1 : units α) :=
+  univ.prod (λ x, x) = (-1 : units K) :=
 begin
   classical,
-  have : ((@univ (units α) _).erase (-1)).prod (λ x, x) = 1,
+  have : ((@univ (units K) _).erase (-1)).prod (λ x, x) = 1,
   from prod_involution (λ x _, x⁻¹) (by simp)
     (λ a, by simp [units.inv_eq_self_iff] {contextual := tt})
     (λ a, by simp [@inv_eq_iff_inv_eq _ _ a, eq_comm] {contextual := tt})
     (by simp),
-  rw [← insert_erase (mem_univ (-1 : units α)), prod_insert (not_mem_erase _ _),
+  rw [← insert_erase (mem_univ (-1 : units K)), prod_insert (not_mem_erase _ _),
       this, mul_one]
 end
 
-end
-
-lemma pow_card_sub_one_eq_one [field α] [fintype α] (a : α) (ha : a ≠ 0) :
-  a ^ (fintype.card α - 1) = 1 :=
-calc a ^ (fintype.card α - 1) = (units.mk0 a ha ^ (fintype.card α - 1) : units α) :
+/-- In a finite field of cardinality `q`, one has `a^(q-1) = 1` for all nonzero `a`. -/
+lemma pow_card_sub_one_eq_one  (a : K) (ha : a ≠ 0) :
+  a ^ (q - 1) = 1 :=
+calc a ^ (q - 1) = (units.mk0 a ha ^ (q - 1) : units K) :
     by rw [units.coe_pow, units.coe_mk0]
   ... = 1 : by { classical, rw [← card_units, pow_card_eq_one], refl }
 
+variable (K)
+
+theorem card (p : ℕ) [char_p K p] : ∃ (n : ℕ+), nat.prime p ∧ q = p^(n : ℕ) :=
+begin
+  haveI hp : fact p.prime := char_p.char_is_prime K p,
+  have V : vector_space (zmod p) K, from { .. (zmod.cast_hom p K).to_module },
+  obtain ⟨n, h⟩ := @vector_space.card_fintype _ _ _ _ V _ _,
+  rw zmod.card at h,
+  refine ⟨⟨n, _⟩, hp, h⟩,
+  apply or.resolve_left (nat.eq_zero_or_pos n),
+  rintro rfl,
+  rw nat.pow_zero at h,
+  have : (0 : K) = 1, { apply fintype.card_le_one_iff.mp (le_of_eq h) },
+  exact absurd this zero_ne_one,
+end
+
+theorem card' : ∃ (p : ℕ) (n : ℕ+), nat.prime p ∧ q = p^(n : ℕ) :=
+let ⟨p, hc⟩ := char_p.exists K in ⟨p, @finite_field.card K _ _ p hc⟩
+
+@[simp] lemma cast_card_eq_zero : (q : K) = 0 :=
+begin
+  rcases char_p.exists K with ⟨p, _char_p⟩, resetI,
+  rcases card K p with ⟨n, hp, hn⟩,
+  simp only [char_p.cast_eq_zero_iff K p, hn],
+  conv { congr, rw [← nat.pow_one p] },
+  exact nat.pow_dvd_pow _ n.2,
+end
+
+/-- The sum of `x^i` as `x` ranges over the units of a finite field of cardinality `q`
+is equal to `0` unless `(q-1) ∣ i`, in which case the sum is `q-1`. -/
+lemma sum_pow_units (i : ℕ) :
```

Comment:
Could this be split up into a lemma about the sum over any nontrivial finite subgroup of the units of a (potentially infinite) integral domain? And then applied to this particular group. The general theorem sounds like it might be useful.

Reply:
I've done this.

Reply:
I was thinking something more like this.
```lean
lemma sum_units_subgroup (f : G →* R) (i : ℕ) :
  ∑ g : G, f g = if f = 1 then fintype.card G else 0
```
This is more general and should easily imply the thing you proved.

Reply:
I think this should be done in a later refactor. It is not straightforward to prove this lemma with current mathlib, I think.

Reply:
This has now been done.